### PR TITLE
Don't separate fully-qualified as a regular `dot qualified expression`

### LINKED ISSUE #1403: Add analysis of fully-qualified expression in NewLines rule
## Describe the bug

Fully-qualified expression should not be separated

## Initial code:

```kotlin
private val holder: java.util.concurrent.atomic.AtomicReference<T> = java.util.concurrent.atomic.AtomicReference(valueToStore)
```

```kotlin
private val holder: kotlin.native.concurrent.AtomicReference<T> = kotlin.native.concurrent.AtomicReference(valueToStore)
```

## Expected behavior

```kotlin
private val holder: java.util.concurrent.atomic.AtomicReference<T> = java.util.concurrent.atomic.AtomicReference(valueToStore)
```

```kotlin
private val holder: kotlin.native.concurrent.AtomicReference<T> = kotlin.native.concurrent.AtomicReference(valueToStore)
```

## Observed behavior

```kotlin
private val holder: java.util.concurrent.atomic.AtomicReference<T> = java.util
        .concurrent
        .atomic
        .AtomicReference(valueToStore)
```

```kotlin
private val holder: kotlin.native.concurrent.AtomicReference<T> = kotlin.native
        .concurrent
        .AtomicReference(valueToStore)
```

## Environment information
* diktat version: 1.2.0
* kotlin version: 1.6

### PR DESCRIPTION
**Pull request by Arrgentum**
## Whats added

 Add analytic fully qualified - split long `dot qualified expression` and `safe access expression` and and their combinations after first `call expression`
 
## Actions checklist
* [ ] correct Warning message
* [ ] Added tests on fixers - 5 last test in `long dot qualified expression`

### CODE REVIEW

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt [hunk 1]
```diff
@@ -26,7 +26,7 @@ class NewlinesRuleWarnTest : LintTestBase(::NewlinesRule) {
             mapOf("maxCallsInOneLine" to "1"))
     )
     private val ruleId = "$DIKTAT_RULE_SET_ID:${NewlinesRule.NAME_ID}"
-    private val dotQuaOrSafeAccessOrPostfixExpression = "${WRONG_NEWLINES.warnText()} should be split before second and other dot/safe access"
+    private val dotQuaOrSafeAccessOrPostfixExpression = "${WRONG_NEWLINES.warnText()} should be split before second and other dot/safe access after first call expression"
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt:29]: I think we shouldn't add this to warning text. This part makes text too complex, and also we don't (or at least shouldn't) forbid splitting these expressions. The main reason in our logic with splitting only after call expression is that we can't distinguish fully-qualified names from chain of property access expressions. I.e., this is still valid:
```
project
    .projectDir
    .absolutePath
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt [hunk 2]
```diff
@@ -193,7 +196,8 @@ class NewlinesRule(configRules: List<RulesConfig>) : DiktatRule(
         list.forEachIndexed { index, astNode ->
             val dotNode = astNode.getFirstChildWithType(DOT) ?: astNode.getFirstChildWithType(SAFE_ACCESS)
             val nodeBeforeDot = dotNode?.treePrev
-            if (index > 0) {
+            val firstElem = astNode.firstChildNode
+            if (index > 0 && (firstElem.textContains('(') || firstElem.textContains('{'))) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt:200]: Expression `firstElem.textContains('(') || firstElem.textContains('{')` occurs 3 times in this class already. It's time to extract it into a function

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt [hunk 3]
```diff
@@ -159,15 +159,17 @@ class NewlinesRule(configRules: List<RulesConfig>) : DiktatRule(
             }.reversed()
             if (listDot.size > 3) {
                 val without = listDot.filterIndexed { index, it ->
-                    val whiteSpaceBeforeDotOrSafeAccess = it.findChildByType(DOT)?.treePrev ?: it.findChildByType(SAFE_ACCESS)?.treePrev
+                    val nodeBeforeDotOrSafeAccess = it.findChildByType(DOT)?.treePrev ?: it.findChildByType(SAFE_ACCESS)?.treePrev
                     val firstElem = it.firstChildNode
-                    (firstElem.textContains('(') || firstElem.textContains('{')) && (index > 0) && ((whiteSpaceBeforeDotOrSafeAccess?.elementType != WHITE_SPACE) ||
-                        (whiteSpaceBeforeDotOrSafeAccess.elementType != WHITE_SPACE && !whiteSpaceBeforeDotOrSafeAccess.textContains('\n')))
+                    val isTextContainsParenthesized = firstElem.textContains('(') || firstElem.textContains('{')
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt:164]:
```suggestion
                    val isTextContainsFunctionCall = firstElem.textContains('(') || firstElem.textContains('{')
```

Comment (Arrgentum) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt]: done

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt (hunk 2 above)
Comment (Arrgentum) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt]: done

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt [hunk 4]
```diff
@@ -158,12 +158,15 @@ class NewlinesRule(configRules: List<RulesConfig>) : DiktatRule(
                 isDotQuaOrSafeAccessOrPostfixExpression(it) && it.elementType != POSTFIX_EXPRESSION
             }.reversed()
             if (listDot.size > 3) {
-                val without = listDot.filterNot {
+                val without = listDot.filterIndexed { index, it ->
                     val whiteSpaceBeforeDotOrSafeAccess = it.findChildByType(DOT)?.treePrev ?: it.findChildByType(SAFE_ACCESS)?.treePrev
-                    whiteSpaceBeforeDotOrSafeAccess?.elementType == WHITE_SPACE && whiteSpaceBeforeDotOrSafeAccess.text.lines().size > 1
+                    val firstElem = it.firstChildNode
+                    (firstElem.textContains('(') || firstElem.textContains('{')) && (index > 0) && ((whiteSpaceBeforeDotOrSafeAccess?.elementType != WHITE_SPACE) ||
+                        (whiteSpaceBeforeDotOrSafeAccess.elementType != WHITE_SPACE && !whiteSpaceBeforeDotOrSafeAccess.textContains('\n')))
```
[the remark this answers is not in the document]
Comment (Arrgentum) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt]: done

**Review by petertrr — APPROVED** [review]

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt (hunk 1 above)
Comment (Arrgentum) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt]: done

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt [hunk 5]
```diff
@@ -624,12 +627,14 @@ class NewlinesRule(configRules: List<RulesConfig>) : DiktatRule(
...
         return if (dropLeadingProperties) {
             // fixme: we can't distinguish fully qualified names (like java.lang) from chain of property accesses (like list.size) for now
-            parentExpressionList?.dropWhile { !it.treeParent.textContains('(') && !it.treeParent.textContains('{') }
+            parentExpressionList?.dropWhile { !isTextContainsFunctionCall(it.treeParent) }
         } else {
             parentExpressionList
         }
     }
 
+    private fun isTextContainsFunctionCall(node: ASTNode): Boolean = node.textContains('(') || node.textContains('}')
```
Comment (nulls) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt:636]: is it expected `}`, not `{` or `)`?

Comment (Arrgentum) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt]: Done

**Pull request merged**